i2c/imx6ull/libimx6ull-i2c.c

```diff
@@ -0,0 +1,361 @@
+	pctl.iopad.sre = 0;
+
+	platformctl(&pctl);
+
+	pctl.iopad.pad = pctl_pad_gpio1_03;
+	platformctl(&pctl);
+#endif
+}
+
+
+static int is_bus_busy(void)
+{
+	return *(i2c.base + i2sr) & (1 << 5);
+}
+
+
+static int wait_bus_busy(int for_busy)
+{
+	int timeout = 10; /* x 10 us = 100us */
+
+	while (!!is_bus_busy() != !!for_busy) {
+		usleep(10);
+		if (--timeout <= 0)
+			return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+
+static int i2c_intr(unsigned int intr, void *data)
+{
+	/* clear interrupt */
+	*(i2c.base + i2sr) &= ~(1 << 1);
+
+	return i2c.cond;
+}
+
+
+static int i2c_trx_complete(void)
+{
+	const time_t timeout_us = 100; /* 100 us */
+	int ret = 0;
+
+	mutexLock(i2c.lock);
+
+	/* wait for transfer complete interrupt - with timeout */
+	while (!(*(i2c.base + i2sr) & (1 << 7))) {
+		if ((ret = condWait(i2c.cond, i2c.lock, timeout_us)) == -ETIME) {
+			ret = -ETIMEDOUT;
+			break;
+		}
+	}
+
+	/* clear interrupt second time - sticky conditionals */
+	*(i2c.base + i2sr) &= ~(1 << 1);
+
+	mutexUnlock(i2c.lock);
+	return ret;
+}
+
+
+static int i2c_acked(void)
+{
+	if (*(i2c.base + i2sr) & (1 << 0))
+		return -EIO;
+
+	return 0;
+}
+
+
+/* writes 1 byte and waits for completion / ack */
+static int write_byte(uint8_t byte)
+{
+	int ret = 0;
+
+	*(i2c.base + i2sr) = 0; /* clear previous interrupts */
+	*(i2c.base + i2dr) = byte;
+
+	/* wait for byte to be sent */
+	ret = i2c_trx_complete();
+	if (ret < 0)
+		return ret;
+
+	ret = i2c_acked();
+	return ret;
+}
+
+
+/* Performs i2c generic write operation to the given slave device. */
+int i2c_busWrite(uint8_t dev_addr, const uint8_t *data, uint32_t len)
+{
+	unsigned int i;
+	int ret;
+
+	/* single-master mode, bus should always be idle when we're not transferring */
+	if (is_bus_busy())
+		return -EBUSY;
+
+	/* generate START: enable master mode, enable TX, enable interrupt */
+	*(i2c.base + i2cr) |= (1 << 6) | (1 << 5) | (1 << 4);
+
+	/* wait for the bus to be busy */
+	ret = wait_bus_busy(1);
+
+	/* send first byte - DEVICE WRITE request */
+	if (ret == 0)
+		ret = write_byte((dev_addr << 1) | 0x0);
+
+	/* send data */
+	for (i = 0; (ret == 0) && (i < len); ++i) {
+		ret = write_byte(data[i]);
+	}
+
+	/* generate STOP: disable master mode, disable TX */
+	*(i2c.base + i2cr) &= ~((1 << 6) | (1 << 5) | (1 << 4));
+
+	/* wait for the bus to be idle */
+	wait_bus_busy(0); /* don't overwrite possible error in ret */
+
+	return ret;
+}
+
+
+/* Performs i2c generic read operation from the given slave device. */
+int i2c_busRead(uint8_t dev_addr, uint8_t *data_out, uint32_t len)
+{
+	unsigned int i;
+	uint16_t temp;
+	int ret;
+
+	/* single-master mode, bus should always be idle when we're not transferring */
+	if (is_bus_busy())
+		return -EBUSY;
+
+	/* generate START: enable master mode, enable TX for sending register address */
+	*(i2c.base + i2cr) |= (1 << 6) | (1 << 5) | (1 << 4);
+
+	/* wait for the bus to be busy */
+	ret = wait_bus_busy(1);
+	if (ret < 0)
+		goto generate_stop;
```

Comment:
`goto`, for better or for worse, is forbidden by our coding guidelines.

Reply:
Yeah, that's just because somebody didn't read the Dijkstra article thoughtfully, but I won't argue with You on that here.

I considered other possibilities - I'm against empty `do-while(0)` as it's only a syntactic sugar over goto on top of increasing cyclomatic complexity and also being advisory rule in MISRA:2012 (_15.4 There should be no more than one break statement used to terminate any iteration statement_, so in the same basket as _15.1 Goto statement should not be used_).

I tried moving body of the function to the separate sub-function but it seemed less readable so I settled on using `goto` for "error handling" causes - which actually is consistent with the allowed usage in linux kernel **and implicitly allowed / proposed by Dijkstra himself** :)
![Screenshot_20210923_112959](https://user-images.githubusercontent.com/3939983/134484857-bc52c2a8-8818-49e7-a181-3adedb450807.png)


I'll go for putting body of the function into separate sub-function but it would be less readable that way,